vm/hv1/hv1_emulator/src/hv.rs

```diff
@@ -436,50 +428,37 @@ impl ProcessorVtlHv {
-                tracelimit::warn_ratelimited!(
-                    error = &err as &dyn std::error::Error,
-                    "failed to clear lazy eoi from assist page"
-                );
-            }
+            self.vp_assist_page[offset..offset + 4].atomic_write_obj(&v);
             false
         }
     }
 
     /// Get the register values to restore on vtl return
-    pub fn return_registers(&self) -> Result<[u64; 2], GuestMemoryError> {
-        let gpa = (self.vp_assist_page.gpa_page_number() * HV_PAGE_SIZE)
-            + offset_of!(hvdef::HvVpAssistPage, vtl_control) as u64
-            + offset_of!(HvVpVtlControl, registers) as u64;
-
-        self.guest_memory.read_plain(gpa)
+    pub fn return_registers(&self) -> [u64; 2] {
+        let offset =
+            offset_of!(hvdef::HvVpAssistPage, vtl_control) + offset_of!(HvVpVtlControl, registers);
+        self.vp_assist_page[offset..offset + 16].atomic_read_obj()
     }
 
     /// Set the reason for the vtl return into the vp assist page
-    pub fn set_return_reason(&self, reason: HvVtlEntryReason) -> Result<(), GuestMemoryError> {
-        let gpa = (self.vp_assist_page.gpa_page_number() * HV_PAGE_SIZE)
-            + offset_of!(hvdef::HvVpAssistPage, vtl_control) as u64
-            + offset_of!(HvVpVtlControl, entry_reason) as u64;
-
-        self.guest_memory.write_plain(gpa, &(reason.0))
+    pub fn set_return_reason(&mut self, reason: HvVtlEntryReason) {
+        let offset = offset_of!(hvdef::HvVpAssistPage, vtl_control)
+            + offset_of!(HvVpVtlControl, entry_reason);
+        self.vp_assist_page[offset..offset + 4].atomic_write_obj(&reason);
     }
 
     /// Gets whether VINA is currently asserted.
-    pub fn vina_asserted(&self) -> Result<bool, GuestMemoryError> {
-        let gpa = (self.vp_assist_page.gpa_page_number() * HV_PAGE_SIZE)
-            + offset_of!(hvdef::HvVpAssistPage, vtl_control) as u64
-            + offset_of!(HvVpVtlControl, vina_status) as u64;
-
-        self.guest_memory.read_plain(gpa).map(|v: u8| v != 0)
+    pub fn vina_asserted(&self) -> bool {
+        let offset = offset_of!(hvdef::HvVpAssistPage, vtl_control)
+            + offset_of!(HvVpVtlControl, vina_status);
+        self.vp_assist_page[offset..offset + 1].atomic_read_obj::<u8>() != 0
     }
 
     /// Sets whether VINA is currently asserted.
-    pub fn set_vina_asserted(&self, value: bool) -> Result<(), GuestMemoryError> {
-        let gpa = (self.vp_assist_page.gpa_page_number() * HV_PAGE_SIZE)
-            + offset_of!(hvdef::HvVpAssistPage, vtl_control) as u64
-            + offset_of!(HvVpVtlControl, vina_status) as u64;
-
-        self.guest_memory.write_plain(gpa, &(value as u8))
+    pub fn set_vina_asserted(&mut self, value: bool) {
+        let offset = offset_of!(hvdef::HvVpAssistPage, vtl_control)
+            + offset_of!(HvVpVtlControl, vina_status);
+        self.vp_assist_page[offset..offset + 1].atomic_write_obj(&(value as u8));
```

Comment:
You can probably just say `self.vp_assist_page[offset].store(value as u8, Relaxed)`